coin_chk_libhdr.m4

```diff
@@ -0,0 +1,432 @@
+   [no],[m4_if([$1],[link],[$4],[no])],
+   [sep],[m4_if([$1],[link],[sep],[$3])],
+   [tog],[m4_if([$1],[link],[tog],[$3])],
+   [m4_if([$1],[link],[$4],[$3])])])dnl 	# COIN_LIBHDR_DFLT_HACK
+  
+
+
+###########################################################################
+#                         FIND_PRIM_LIBHDR                                #
+###########################################################################
+
+# COIN_FIND_PRIM_LIBHDR([prim],[lflgs],[cflgs],[dflgs],
+#                       [function-body],[includes],
+#                       [dfltaction],[cmdlineopts])
+
+# Determine whether we can use primary library prim ($1) and assemble
+# information on the required linker flags (prim_lflags), compiler flags
+# (prim_cflags), and data directories (prim_data) as specified by cmdlineopts
+# ($8).
+
+# A compile check will be performed if [includes] ($6) is specified.  A link
+# check will be performed if [function-body] ($5) is specified. The default is
+# to use a program composed by concatenating [includes] and [function-body]
+# with link flags specified by [lflgs] ($2), and compiler flags specified by
+# [cflgs] ($3). If [dfltaction] ($7) ends in 'sep', the link check will be
+# performed without the header (see LIBHDR_DFLT_HACK).
+
+# There's really no reasonable assumptions that can be made about how a data
+# directory is specified, so there's no attempt to test for existence.
+
+# [cmdlineopts] ($8) specifies the set of configure command line options
+# processed: 'nodata' processes --with-prim, --with-prim-lflags,
+# and --with-prim-cflags; 'dataonly' processes only --with-prim and
+# --with-prim-data; anything else ('all' works well) processes all four
+# command line options. Shell code produced by the macro is tailored based
+# on [cmdlineopts]. `nodata' is the default.
+
+# --with-prim is interpreted as follows:
+#   * --with-prim=no is equivalent to --without-prim
+#     prim status is set to skipping
+#   * --with-prim or --with-prim=yes is equivalent to
+#       --with-prim-lflags=-lprim
+#       --with-prim-data=/usr/local/share/prim
+#     if the user doesn't override lflags & data from the command line
+#     prim status is set to requested
+#   * Any other value is taken as equivalent to
+#       --with-prim-data=value (dataonly) or
+#       --with-prim-lflags=value (anything else)
+#     prim status is set to requested
+
+# The algorithm first checks for a user-specified value of --with-prim; values
+# are interpreted as above.  Next, it looks for user specified values given
+# with command line parameters --with-prim-lflags, --with-prim-cflags, and
+# --with-prim-data. If any of these are specified, the value overrides the
+# value passed as a parameter.
+
+# The usage portion of [dfltaction] ($7) (no, yes) is used as the default
+# value of --with-prim if the user offers no guidance via command line
+# parameters. The (hardwired) default is yes.  See LIBHDR_DFLT_HACK for the
+# full story.
```

Comment:
I doubt I fully understood the whole story about separate compile and link checks.

But I wonder whether one couldn't infer what to do from whether function-body and/or includes is empty.
- If a function-body is given and is not only a single word, assemble a code that compiles and links from both includes and function-body.
- If function-body is a single word, then do a linkcheck only. If includes are given, raise an error?
- If only includes are given, then check whether that compiles with an empty function-body.

Reply:
I'm going to argue your philosophical position: don't guess.  Also, the macros are set up in such a way that most users will never need to bother with `[dfltaction]`. The macros assume that if you're asking about a library, the default should be to use it; overriding this requires only `no` as `[dfltaction]`.  Explicitly specifying the link test strategy (`sep` or `tog`) should hardly ever be necessary.

The `CHK_LIB` wrapper macro is defined to handle the common case where the user supplies only a function (i.e., single word) and a header file name and wants separate checks for header usability (check that `#include header` compiles) and ability to link (`void function()` can be linked). It hides the complication of the link test strategy from the user by specifying `sep` but won't override an explicit user spec if it's present in `[dfltaction]` (even though forcing `tog` is almost certainly an error here).

The underlying `CHK_LIBHDR` already does the right thing when either of `[function-body]` or `[includes]` are null. The only ambiguity is when both are non-null. `CHK_LIBHDR` assumes that if both are present, they should be tested as a whole program in the link check. The only time the user needs to intervene is if they're calling `CHK_LIBHDR` directly, both `[function-body]` and `[includes]` are non-null, and they really want separate compile and link tests. (As per above, this is exactly the case that `CHK_LIB` hides from the user.)

Reply:
Could one then at least write out the name of the strategies? `sep` I can still guess, but that `tog` should be short for `together` is not so intuitive. If one reads a configure.ac file, it would be good if one can get some idea what these flags do without having to read up the coin.m4 file and deciphering LIBHDR_DFLT_HACK.

Reply:
Maybe something like {yes, no} x {separate, together} with a space in between for readability (e.g., `yes separate`). Full words for sure. I'll see if the space works.

Reply:
Check the new version of `LIBHDR_DFLT_HACK`. It parses strings of space-separated keywords. I made the keywords more human-friendly. As collateral damage, you can now write keywords in any order. Should be easily expansible to handle more overload types and keywords, as long as the keywords are unique.  What we might want to do is change the `yes` and `no` keywords for usage to something like `default_use` and `default_skip`. Might break existing configure.ac, but the `[dfltaction]` parameter of the old `CHK_LIB` seems to be rarely used. No examples that I see in the basic optimisation stack. Any that you know of?

Reply:
That would be a nice change.

I don't remember a place where it is currently used. Usually one wants to use everything there is available.

Reply:
Done, with the latest commits.